src/stan/mcmc/chainset.hpp

```diff
@@ -0,0 +1,567 @@
+   * @return pair (bulk_ess, tail_ess)
+   */
+  std::pair<double, double> split_rank_normalized_ess(
+      const std::string& name) const {
+    return split_rank_normalized_ess(index(name));
+  }
+
+  /**
+   * Computes the mean Monte Carlo error estimate for the central 90% interval.
+   * See https://arxiv.org/abs/1903.08008, section 4.4.
+   * Follows implementation in the R posterior package
+   *
+   * @param index parameter index
+   * @return pair (bulk_ess, tail_ess)
+   */
+  double mcse_mean(const int index) const {
+    double ess_bulk = analyze::split_rank_normalized_ess(samples(index)).first;
+    return sd(index) / std::sqrt(ess_bulk);
+  }
+
+  /**
+   * Computes the mean Monte Carlo error estimate for the central 90% interval.
+   * See https://arxiv.org/abs/1903.08008, section 4.4.
+   * Follows implementation in the R posterior package.
+   *
+   * @param name parameter name
+   * @return pair (bulk_ess, tail_ess)
+   */
+  double mcse_mean(const std::string& name) const {
+    return mcse_mean(index(name));
+  }
+
+  /**
+   * Computes the standard deviation of the Monte Carlo error estimate
+   * https://arxiv.org/abs/1903.08008, section 4.4.
+   * Follows implementation in the R posterior package.
+   *
+   * @param index parameter index
+   * @return pair (bulk_ess, tail_ess)
+   */
+  double mcse_sd(const int index) const {
+    Eigen::MatrixXd s = samples(index);
+    Eigen::MatrixXd s2 = s.array().square();
+    double ess_s = analyze::split_rank_normalized_ess(s).first;
+    double ess_s2 = analyze::split_rank_normalized_ess(s2).first;
+    double ess_sd = std::min(ess_s, ess_s2);
+    return sd(index)
+           * std::sqrt(std::exp(1) * std::pow(1 - 1 / ess_sd, ess_sd - 1) - 1);
+  }
+
+  /**
+   * Computes the standard deviation of the Monte Carlo error estimate
+   * https://arxiv.org/abs/1903.08008, section 4.4.
+   * Follows implementation in the R posterior package
+   *
+   * @param name parameter name
+   * @return pair (bulk_ess, tail_ess)
+   */
+  double mcse_sd(const std::string& name) const { return mcse_sd(index(name)); }
+
+  /**
+   * Compute autocorrelation for one column of one chain.
+   * Throws exception if column index is out of bounds.
+   * Autocorrelation is computed using Stan math library implmentation.
+   *
+   * @param chain chain index
+   * @param index column index
+   * @return vector of chain autocorrelation at all lags
+   */
+  Eigen::VectorXd autocorrelation(const int chain, const int index) const {
+    if (chain < 0 || chain >= num_chains()) {
+      std::stringstream ss;
+      ss << "Bad index " << index << ", should be between 0 and "
+         << (num_chains() - 1);
+      throw std::invalid_argument(ss.str());
+    }
+    if (index < 0 || index >= param_names().size()) {
+      std::stringstream ss;
+      ss << "Bad index " << index << ", should be between 0 and "
+         << (num_params() - 1);
+      throw std::invalid_argument(ss.str());
+    }
```

Comment:
Rest of this file is not doing bounds checking -- either behavior seems fine but should be consistent

Reply:
gone.